# Final Project - Lập trình cho Khoa học dữ liệu

## Thông tin nhóm

Nhóm: 12

Số lượng thành viên: 2

Thành viên 1:
- Họ và tên:
- MSSV:
- Email:

Thành viên 2:
- Họ và tên:
- MSSV:
- Email:

## Thông tin đồ án

## Các thư viện hỗ trợ

In [1]:
import pandas as pd

## Thu thập dữ liệu
> What subject is your data about? What is the source of your data?

Đây là tập dữ liệu xuyên quốc gia chứa tất cả các giao dịch diễn ra từ ngày 01/12/2010 đến 09/12/2011 cho một cửa hàng bán lẻ trực tuyến (không có cửa hàng trực tiếp) và có trụ sở tại Vương quốc Anh. Công ty chủ yếu bán quà tặng các dịp lễ. Đa số khách hàng của công ty là những thương lái.

> Do authors of this data allow you to use like this?

Không có thông tin về License trên Kaggle, tuy nhiên tập dữ liệu được hiển thị public trên Kaggle và được lấy từ một nguồn online miễn phí khác ([UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/index.php))

> How did authors collect data?

Không có thông tin về việc tác giả thu thập như thế nào, tuy nhiên người upload tập dữ liệu này trên Kaggle, lấy trực tiếp từ [UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/index.php) và sở hữu thực tế được ghi trên thư viện này là "*Dr Daqing Chen, Director: Public Analytics group. chend '@' lsbu.ac.uk, School of Engineering, London South Bank University, London SE1 0AA, UK*".



In [2]:
path = "./data.csv"
content = pd.read_csv(path, encoding= 'unicode_escape')
content

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

## Khám phá dữ liệu

In [3]:
# Trình bày mỗi câu hỏi bằng một cell markdown, định dạng h3 (### ở đầu).
# Đặt câu hỏi trước khi thực hiện trả lời câu hỏi (bằng code hoặc text).

### Dữ liệu có bao nhiêu dòng và bao nhiêu cột?

In [4]:
rows = content.shape[0]
cols = content.shape[1]
print(f"Dữ liệu trên có {rows} dòng và {cols} cột")

Dữ liệu trên có 541909 dòng và 8 cột


### Mỗi dòng có ý nghĩa gì? 

Mỗi dòng là thông tin giao dịch một loại hàng hoá của một khách hàng.

### Trong dữ liệu trên  ( chưa qua xử lý), có dòng bị lặp không?

In [5]:
print(f"Dữ liệu trên có {content.duplicated().sum()} dòng bị lặp")

Dữ liệu trên có 5268 dòng bị lặp


### Mỗi cột có ý nghĩa gì?

- <b>InvoiceNo</b>: Mã hoá đơn, , nếu mã này có kí tự đầu là "C" thì giao dịch đó đã bị huỷ.
- <b>StockCode</b>: Mã sản phẩm
- <b>Description</b>: Tên sản phẩm
- <b>Quantity</b>: Số lượng mỗi sản phẩm trên mỗi lần giao dịch
- <b>InvoiceDate</b>: Ngày và giờ giao dịch
- <b>UnitPrice</b>: Đơn giá. Đơn giá này được tính theo đơn vị Bảng Anh.
- <b>CustomerID</b>: Mã khách hàng
- <b>Country</b>: Tên quốc gia của khách hàng

### Mỗi cột đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để xử lý ở những bước tiếp theo không?

In [6]:
pd.DataFrame(content.dtypes).T.rename(index={0:'type'})

InvoiceNo StockCode Description Quantity InvoiceDate UnitPrice  \
type    object    object      object    int64      object   float64   

     CustomerID Country  
type    float64  object

- Các cột <b>InvoiceNo</b>, <b>StockCode</b>, <b>Description</b>, <b>Country</b>: Có kiểu dữ liệu ở dạng chuỗi. Kiểu dữ liệu này thích hợp cho các loại dữ liệu trên nên đây là kiểu dữ liệu đúng.
- Các cột <b>Quantity</b>, <b>CustomerID</b>, <b>UnitPrice</b>: Có kiểu dữ liệu dạng số. Thể hiện đúng tên của từng cột.
- Cột <b>InvoiceDate</b>: Có kiểu dữ liệu dạng chuỗi. Để thuận tiện hơn khi xử lý thời gian, cần chuyển về dạng datetime. Em sẽ chuyển ngay dưới đây.

In [7]:
content['InvoiceDate'] = pd.to_datetime(content['InvoiceDate'])
categorical_cols = ['InvoiceNo', 'StockCode', 'Description', 'Country']
numeric_cols = ['Quantity', 'CustomerID', 'UnitPrice']

### Với mỗi cột dạng số thực và dạng datetime, các giá trị được phân bố như thế nào?

Ở đây có 3 cột số thực là <b>Quantity</b>, <b>CustomerID</b> và <b>UnitPrice</b>. Em sẽ khám phá xem tỷ lệ % giá trị bị thiếu của các cột trên.

In [8]:
null_numeric = pd.DataFrame(pd.DataFrame(content[numeric_cols].isnull().sum()/content.shape[0]*100).T.
                         rename(index={0:'Missing (%)'}))
null_numeric

Quantity  CustomerID  UnitPrice
Missing (%)       0.0   24.926694        0.0

Nhìn vào bảng trên, nhận thấy cột <b>Quantity</b> và cột <b>UnitPrice</b> có 0% tỷ lệ thiếu. Nhưng cột <b>CustomerID</b> có đến 24.9% tỷ lệ thiếu. Đây là cột CustomerID, mỗi khách hàng có một CustomerID duy nhất, nhóm em nhận thấy không thể dùng những kỹ thuật điền giá trị thiếu để điền vào trong trường hợp này được. Do đó, em sẽ xoá những dòng có CustomerID bị thiếu.

In [9]:
content['InvoiceDate'] = pd.to_datetime(content['InvoiceDate'])
min_Quantity = min(content['Quantity'])
max_Quantity = max(content['Quantity'])
min_UnitPrice = min(content['UnitPrice'])
max_UnitPrice = max(content['UnitPrice'])
min_InvoiceDate = min(content['InvoiceDate']) 
max_InvoiceDate = max(content['InvoiceDate'])
data_min_max = {
    'Quantity': [min_Quantity, max_Quantity],
    'UnitPrice': [min_UnitPrice, max_UnitPrice],
    'InvoiceDate': [min_InvoiceDate, max_InvoiceDate]
}
min_max_numeric = pd.DataFrame(data_min_max).rename(index={0:'min', 1:'max'})
min_max_numeric

Quantity  UnitPrice         InvoiceDate
min    -80995  -11062.06 2010-12-01 08:26:00
max     80995   38970.00 2011-12-09 12:50:00

min và max của cột <b>InvoiceDate</b> không có gì bất thường. Nhưng min và max của cột <b>Quantity</b> và cột <b>UnitPrice</b> chứa số âm. Số âm này thể hiện số lượng trả lại và giá trị đơn trả lại. Nên đây cũng không có gì bất thường.

### Với mỗi cột có kiểu dữ liệu dạng phân loại, các giá trị được phân bố như thế nào?

In [10]:
def dif_values(arr):
    """
    Đầu vào:
      arr: 
        Dạng: list 
        Miêu tả: Dữ liệu của một cột
    
    Đầu ra:
      set_dif_val:
        Miêu tả: Trả về một tập chứa các giá trị khác nhau có trong list
      num_of_dif: 
        Miêu tả: Trả về số lượng các giá trị khác nhau có trong list
      """
    set_dif_val = set(arr)
    set_dif_val = [x for x in set_dif_val if x==x]
    num_of_dif = len(set_dif_val)
    
    return set_dif_val, num_of_dif

In [16]:
categorical_info = {}
num_of_missing_categorical = pd.Series(content[categorical_cols].isnull().sum())
persent_missing_categorical = pd.Series(content[categorical_cols].isnull().sum()/content.shape[0]*100)
for col in categorical_cols:
    categorical_info[col] = dif_values(content[col])
print(f"{'Categorical':15} {'Miss(%)':12} {'Num_of_diff':16} {'Some_val'}")

for col in categorical_cols:
    categorical_info['StockCode'][0][:3]
    print(f'{col:15} {persent_missing_categorical[col]:<12.3f} {categorical_info[col][1]:<12} {categorical_info[col][0].__repr__()[:40]+"...":<}')

Categorical     Miss(%)      Num_of_diff      Some_val
InvoiceNo       0.000        25900        ['C574443', '571628', '547967', 'C566758...
StockCode       0.000        4070         ['22813', '22100', '23061', '62074B', '2...
Description     0.268        4223         ['CHEST OF DRAWERS GINGHAM HEART ', 'EMP...
Country         0.000        38           ['Hong Kong', 'Denmark', 'Switzerland', ...


Theo bảng trên, chỉ có cột Description có dữ liệu bị thiếu, và thiếu 0.268%. Mặc dù có 4070 mã sản phẩm nhưng lại có tới 4223 tên cho sản phẩm. Số lượng các giá trị khác nhau trong cột dạng categorical đều là số dương, cột some_val đều khác nhau. Dữ liệu thu thập được không có gì bất thường. Để có cái nhìn rõ hơn về các giá trị còn thiếu, nhóm em sẽ in ra các dòng còn thiếu.

In [48]:
des_missing = content[content['Description'].isna()]
des_missing

InvoiceNo StockCode Description  Quantity         InvoiceDate  \
622       536414     22139         NaN        56 2010-12-01 11:52:00   
1970      536545     21134         NaN         1 2010-12-01 14:32:00   
1971      536546     22145         NaN         1 2010-12-01 14:33:00   
1972      536547     37509         NaN         1 2010-12-01 14:33:00   
1987      536549    85226A         NaN         1 2010-12-01 14:34:00   
...          ...       ...         ...       ...                 ...   
535322    581199     84581         NaN        -2 2011-12-07 18:26:00   
535326    581203     23406         NaN        15 2011-12-07 18:31:00   
535332    581209     21620         NaN         6 2011-12-07 18:35:00   
536981    581234     72817         NaN        27 2011-12-08 10:33:00   
538554    581408     85175         NaN        20 2011-12-08 14:06:00   

        UnitPrice  CustomerID         Country  
622           0.0         NaN  United Kingdom  
1970          0.0         NaN  United Kingdom  
1971          0.0         NaN  United Kingdom  
1972          0.0         NaN  United Kingdom  
1987          0.0         NaN  United Kingdom  
...           ...         ...             ...  
535322        0.0         NaN  United Kingdom  
535326        0.0         NaN  United Kingdom  
535332        0.0         NaN  United Kingdom  
536981        0.0         NaN  United Kingdom  
538554        0.0         NaN  United Kingdom  

[1454 rows x 8 columns]

In [56]:
max_price = content[content['Description'].isna()]['UnitPrice'].max()
customer_not_missing = sum(content[content['Description'].isna()]['CustomerID'].isna())
print(f"Giá trị lớn nhất khi giao dịch của các giao dịch khi bị thiếu Description là: {max_price}")
print(f"Số lượng CustomerID bị thiếu khi Description bị thiếu là: {customer_not_missing}")

Giá trị lớn nhất khi giao dịch của các giao dịch khi bị thiếu Description là: 0.0
Số lượng CustomerID bị thiếu khi Description bị thiếu là: 1454


Nhìn vào các giá trị ở cột <b>Description</b> bị thiếu, nhận thấy rằng toàn bộ giá trị của cột <b>UnitPrice</b> đều có giá trị = 0 và toàn bộ giá trị cột <b>CustomerID</b> đều bị thiếu. Dữ liệu bất thường do không có CustomerID và UnitPrice = 0.

## Đặt vấn đề

In [12]:
# Trình bày mỗi câu hỏi bằng một cell markdown, định dạng h3 (### ở đầu).
# Đặt câu hỏi trước khi thực hiện trả lời câu hỏi (bằng code hoặc text).

## Tiền xử lí dữ liệu

In [13]:
# Trình bày mỗi câu hỏi bằng một cell markdown, định dạng h3 (### ở đầu).
# Đặt câu hỏi trước khi thực hiện trả lời câu hỏi (bằng code hoặc text).

## Phân tích dữ liệu và trả lời câu hỏi

In [14]:
# Trình bày mỗi câu hỏi bằng một cell markdown, định dạng h3 (### ở đầu).
# Đặt câu hỏi trước khi thực hiện trả lời câu hỏi (bằng code hoặc text).

## Nguồn tham khảo